modules/core/src/matrix.cpp

```diff
@@ -315,7 +315,7 @@ void finalizeHdr(Mat& m)
         m.rows = m.cols = -1;
     if(m.u)
         m.datastart = m.data = m.u->data;
-    if( m.data )
+    if( m.data && d > 0 )
```

Comment:
adding `d > 0` is fine.

I'd also suggest to modify the else branch:

```
else
        m.dataend = m.datalimit = 0;
```

to 

```
else
        m.dataend = m.datalimit = m.data;
```

that will make behaviour of both constructors of an empty matrix more consistent.